z.mlp

```diff
@@ -23,36 +23,194 @@ exception Overflow
+    if (z lxor unsafe_to_int x) land (z lxor unsafe_to_int y) >= 0 
+    then of_int z
+    else c_add x y
+  end else
+    c_add x y
+
+external c_sub: t -> t -> t = "ml_z_sub"
+
+let sub x y =
+  if is_small_int x && is_small_int y then begin
+    let z = unsafe_to_int x - unsafe_to_int y in
+    (* Overflow check -- Hacker's Delight, section 2.12 *)
+    if (unsafe_to_int x lxor unsafe_to_int y)
+       land (z lxor unsafe_to_int x) >= 0 
+    then of_int z
+    else c_sub x y
+  end else
+    c_sub x y
+
+external mul_overflows: int -> int -> bool = "ml_z_mul_overflows" @NOALLOC
+external c_mul: t -> t -> t = "ml_z_mul"
+
+let mul x y =
+  if is_small_int x && is_small_int y
+  && not (mul_overflows (unsafe_to_int x) (unsafe_to_int y))
+  then of_int (unsafe_to_int x * unsafe_to_int y)
+  else c_mul x y
+
+external c_div: t -> t -> t = "ml_z_div"
+
+let div x y =
+  if is_small_int y then
+    if unsafe_to_int y = -1 then
+      neg x
+    else if is_small_int x then
+      of_int (unsafe_to_int x / unsafe_to_int y)
+    else
+      c_div x y
+  else
+    c_div x y
+
 external cdiv: t -> t -> t = "ml_z_cdiv"
 external fdiv: t -> t -> t = "ml_z_fdiv"
-external rem: t -> t -> t = rem@ASM
+
+external c_rem: t -> t -> t = "ml_z_rem"
+
+let rem x y =
+  if is_small_int y then
+    if unsafe_to_int y = -1 then
+      of_int 0
+    else if is_small_int x then
+      of_int (unsafe_to_int x mod unsafe_to_int y)
+    else
+      c_rem x y
+  else
+    c_rem x y
+
 external div_rem: t -> t -> (t * t) = "ml_z_div_rem"
-external succ: t -> t = succ@ASM
-external pred: t -> t = pred@ASM
-external abs: t -> t = abs@ASM
-external logand: t -> t -> t = logand@ASM
-external logor: t -> t -> t = logor@ASM
-external logxor: t -> t -> t = logxor@ASM
-external lognot: t -> t = lognot@ASM
-external shift_left: t -> int -> t = shift_left@ASM
-external shift_right: t -> int -> t = shift_right@ASM
-external shift_right_trunc: t -> int -> t = shift_right_trunc@ASM
-external of_int: int -> t = "ml_z_of_int" @NOALLOC
+
+external c_divexact: t -> t -> t = "ml_z_divexact"
+
+let divexact x y =
+  if is_small_int y then
+    if unsafe_to_int y = -1 then
+      neg x
+    else if is_small_int x then
+      of_int (unsafe_to_int x / unsafe_to_int y)
+    else
+      c_divexact x y
+  else
+    c_divexact x y
+
+external c_succ: t -> t = "ml_z_succ"
+
+let succ x =
+  if is_small_int x && unsafe_to_int x <> max_int
+  then of_int (unsafe_to_int x + 1)
+  else c_succ x
+
+external c_pred: t -> t = "ml_z_pred"
+
+let pred x =
+  if is_small_int x && unsafe_to_int x <> min_int
+  then of_int (unsafe_to_int x - 1)
+  else c_pred x
+
+external c_abs: t -> t = "ml_z_abs"
+
+let abs x =
+  if is_small_int x then
+    if unsafe_to_int x >= 0 then x
+    else if unsafe_to_int x <> min_int then
+      of_int (- unsafe_to_int x)
+    else
+      c_abs x
+  else
+    c_abs x
+
+external c_logand: t -> t -> t = "ml_z_logand"
+
+let logand x y =
```

Comment:
Wouldn't it be possible to preserve the optimization of the original assembly code here?
```
let logand x y =
  let candidate = unsafe_to_int x land unsafe_to_int y in
  if is_small_int candidate
  then of_int candidate
  else c_logand x y
```
(I do not expect the OCaml compiler to find this one, but I expect that the version above is safe?)

Reply:
Thinking about the nature of `candidate` when at least one of `x` or `y` is not an `int`, maybe my proposed version shouldn't be considered safe.

Reply:
You're right, I think this is unsafe, at least in bytecode.  If at least one of x and y is heap-allocated, the result (`candidate`) of the "land" is tagged like a pointer (low bit 0) but points nowhere.  If the GC sees the `candidate` value as a root, it can bomb.  In bytecode, any GC triggered by `c_logand x y` will see `candidate` as a root, because it was pushed on the bytecode interpreter stack.  In native-code, `candidate` is dead in the `else` branch, so the GC should not see it, but Flambda optimizations have been observed to extend the lifetime of local variables, so the GC might see it in the end.